### DISCUSSION TITLE
Is this expected - Self-transition does not cause the same state to be re-entered unless 'reenter: true' is specified for the transition?

### DISCUSSION THREAD

**Discussion by galliumio** [statelyai/xstate#4650] [Q&A]
Apparently this is the expected behavior, as it does re-enter the state as long as the property 'reenter: true" is set for the transition. However, this does not confirm to the SCXML spec (https://www.w3.org/TR/scxml/).

According to the V5 migration guide, 'external transitions previously specified with internal: false are now specified with reenter: true'. In this sense, 'reenter: true' could be regarded as equivalent to an 'external transition' in SCXML.

However in SCXML, the distinction between 'internal' and 'external' transitions does NOT apply to self-transitions. It only applies to transitions to a nested state, according to this statement:

"The behavior of transitions with 'type' of "internal" is identical, except in the case of a transition whose source state is a compound state and whose target(s) is a descendant of the source. In such a case, an internal transition will not exit and re-enter its source state, while an external one will, ..." (Here it is assumed that a descendant state means a 'proper' substate of a compound state.)

In other words, according to SCXML, a state should always be exited and re-entered upon a self-transition to itself regardless to whether it is an internal or external transition (or equivalently whether 'reenter: true' is set).

We can work around this issue by remembering to add 'reenter: true' for self-transitions to the same state. However, this could easily be overlooked and cause unexpected runtime issues when exit/entry actions not called as expected.

**Comment by davidkpiano** [comment]
cc. @Andarist 

We deviated from SCXML here because `internal` vs. `external` was causing too much confusion, and we feel that it's better to be explicit about whether you want a transition to reenter or not. We feel that this is preferable than needing to memorize rules about which transitions are implicitly internal vs. external.
